pep-0544.txt

```diff
@@ -0,0 +1,949 @@
+* ``SupportsAbs`` (and other ``Supports*`` classes)
+* ``Iterable``, ``Iterator``
+* ``Sized``
+* ``Container``
+* ``Collection``
+* ``Reversible``
+* ``Sequence``, ``MutableSequence``
+* ``AbstractSet``, ``MutableSet``
+* ``Mapping``, ``MutableMapping``
+* ``ItemsView`` (and other ``*View`` classes)
+* ``AsyncIterable``, ``AsyncIterator``
+* ``Awaitable``
+* ``Callable``
+* ``ContextManager``, ``AsyncContextManager``
+
+Most of these classes are small and conceptually simple. It is easy to see
+what are the methods these protocols implement, and immediately recognize
+the corresponding runtime protocol counterpart.
+Practically, few changes will be needed in ``typing`` since some of these
+classes already behave the necessary way at runtime. Most of these will need
+to be updated only in the corresponding ``typeshed`` stubs [typeshed]_.
+
+All other concrete generic classes such as ``List``, ``Set``, ``IO``,
+``Deque``, etc are sufficiently complex that it makes sense to keep
+them non-protocols (i.e. require code to be explicit about them). Also, it is
+too easy to leave some methods unimplemented by accident, and explicitly
+marking the subclass relationship allows type checkers to pinpoint the missing
+implementations.
+
+
+Introspection
+-------------
+
+The existing class introspection machinery (``dir``, ``__annotations__`` etc)
+can be used with protocols. In addition, all introspection tools implemented
+in the ``typing`` module will support protocols. Since all attributes need
+to be defined in the class body based on this proposal, protocol classes will
+have even better perspective for introspection than regular classes where
+attributes can be defined implicitly -- protocol attributes can't be
+initialized in ways that are not visible to introspection
+(using ``setattr()``, assignment via ``self``, etc.). Still, some things like
+types of attributes will not be visible at runtime in Python 3.5 and earlier,
+but this looks like a reasonable limitation.
+
+There will be only limited support of ``isinstance()`` and ``issubclass()``
+as discussed above (these will *always* fail with ``TypeError`` for
+subscripted generic protocols, since a reliable answer could not be given
+at runtime in this case). But together with other introspection tools this
+give a reasonable perspective for runtime type checking tools.
+
+
+.. _rejected:
+
+Rejected/Postponed Ideas
+========================
+
+Make every class a protocol by default
+--------------------------------------
+
+Some languages such as Go make structural subtyping the only or the primary
+form of subtyping. We could achieve a similar result by making all classes
+protocols by default (or even always). However we believe that it is better
+to require classes to be explicitly marked as protocols, for the following
+reasons:
+
+* Protocols don't have some properties of regular classes. In particular,
+  ``isinstance()`` is not always well-defined for protocols, whereas it is
```

Comment:
A better argument would be something like this: `isinstance`, as defined for classes that don't explicitly extend `Protocol`, is based on the nominal hierarchy. In order to make everything a protocol by default *and* have `isinstance` work would require changing the semantics of `isinstance`, which won't happen.